stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -414,6 +414,43 @@ function triu(A::Symmetric, k::Integer=0)
     end
 end
 
+function dot(A::Symmetric{T, Matrix{T}}, B::Symmetric{T, Matrix{T}}) where T
+    dotprod = zero(T)
+    m, n = size(A)
+    mB, nB = size(B)
+
+    if m != mB || n != nB
+        throw(DimensionMismatch("A has dimensions ($m,$n) but B has dimensions ($mB,$nB)"))
+    end
+
+    @inbounds if A.uplo == 'U' && B.uplo == 'U'
+        for j in 1:n
+            for i in 1:(j - 1)
+                dotprod += 2 * A.data[i, j] * conj(B.data[i, j])
```

Comment:
Actually, I wonder if it should be
```julia
dotprod += 2real(dot(A.data[i, j], B.data[i, j]))
```
because, for a symmetric block matrix `X` with complex matrix blocks, you would want that `dot(X, X)` returns a real number.

Reply:
Yes, I've noticed this too by looking up other implementation of currently available `dot` methods: They all use a *inner* `dot` method like you suggested. I'm not familiar with block matrices in the Julia standard library yet, but I will look this up.

I guess it will also make sense to add a test for symmetric block matrices.

Reply:
I think the main guideline to follow is to make/keep `dot` recursive, something I also realized only while developing #32739. Whenever you're tempted to write out a product and conjugate one of the factors, just use `dot` instead and you'll get it right. At the very bottom of the hierarchy when it reaches numbers, this is what it's going to do anyway.